### PR TITLE
Support web socket throttling

### LINKED ISSUE #1941: Web socket throttling support
### Describe your problem(s)
Web socket throttling based on the number of frames sent through the websocket connection.

### How will you implement it

---
#### Related Issues:
Sub task of #1540

### PR DESCRIPTION
**Pull request by NomadXD**
### Purpose
Add support for web socket throttling. (API, application and subscription)

### Issues

Fixes #1941

### Automation tests
 - Unit tests added: No
 - Integration tests added: No

### Tested environments

Ubuntu 20.04, JDK 11, envoy-alpine:v1.18.2

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: adapter/internal/api/apis_impl.go [hunk 1]
```diff
@@ -136,7 +136,8 @@ func extractAPIProject(payload []byte) (apiProject ProjectAPI, err error) {
 			}
 			upstreamCerts = append(upstreamCerts, unzippedFileBytes...)
 			upstreamCerts = append(upstreamCerts, newLineByteArray...)
-		} else if strings.Contains(file.Name, apiYAMLFile) || strings.Contains(file.Name, apiJSONFile) {
+		} else if (strings.Contains(file.Name, apiYAMLFile) || strings.Contains(file.Name, apiJSONFile)) &&
```
Comment (NomadXD) [inline adapter/internal/api/apis_impl.go:139]: Extra condition added to differentiate `asyncapi.yaml` from `api.yaml`

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go [hunk 2]
```diff
@@ -497,7 +498,8 @@ func createRoute(params *routeCreateParams) *routev3.Route {
 					},
 					Substitution: endpointBasepath,
 				},
-				UpgradeConfigs: getUpgradeConfig(apiType),
+				UpgradeConfigs:    getUpgradeConfig(apiType),
+				MaxStreamDuration: getMaxStreamDuration(apiType),
```
Comment (NomadXD) [inline adapter/internal/oasparser/envoyconf/routes_with_clusters.go:502]: with envoy 1.18, we have to configure routes for `MaxStreamDuration`. getMaxStreamDuration returns max stream duration for websocket and for http it returns nil (no change for http) https://www.envoyproxy.io/docs/envoy/v1.18.2/version_history/v1.18.0 cc @VirajSalaka

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go [hunk 3]
```diff
@@ -833,3 +835,16 @@ func createAddress(remoteHost string, port uint32) *corev3.Address {
...
 	return &address
 }
+
+// getMaxStreamDuration configures a maximum duration for a websocket route.
+func getMaxStreamDuration(apiType string) *routev3.RouteAction_MaxStreamDuration {
+	var maxStreamDuration *routev3.RouteAction_MaxStreamDuration = nil
+	if apiType == mgw.WS {
+		maxStreamDuration = &routev3.RouteAction_MaxStreamDuration{
+			MaxStreamDuration: &durationpb.Duration{
+				Seconds: 60 * 60 * 24,
```
Comment (NomadXD) [inline adapter/internal/oasparser/envoyconf/routes_with_clusters.go:845]: configured 24 hrs as `MaxStreamDuration`.

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go [hunk 4]
```diff
@@ -810,7 +812,7 @@ func genRouteCreateParams(swagger *model.MgwSwagger, resource *model.Resource, v
 		endpointBasePath:  endpointBasePath,
 		corsPolicy:        swagger.GetCorsConfig(),
 		resourcePathParam: "",
-		resourceMethods:   nil,
+		resourceMethods:   []string{"GET"},
```
Comment (Rajith90) [inline adapter/internal/oasparser/envoyconf/routes_with_clusters.go:815]: Will this add default get resource to REST APIs as well

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go (hunk 3 above)
Comment (Rajith90) [reply adapter/internal/oasparser/envoyconf/routes_with_clusters.go:845]: Is this something we should read from the config

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/api/APIFactory.java [hunk 5]
```diff
@@ -64,10 +66,18 @@ public void addApis(List<Api> apis) {
...
-            RestAPI enforcerApi = new RestAPI();
-            enforcerApi.init(api);
-            String apiKey = getApiKey(enforcerApi);
-            newApis.put(apiKey, enforcerApi);
+            if (api.getApiType().equals(APIConstants.ApiType.REST)) {
+                RestAPI enforcerApi = new RestAPI();
+                enforcerApi.init(api);
+                String apiKey = getApiKey(enforcerApi);
+                newApis.put(apiKey, enforcerApi);
+            } else {
```
Comment (Rajith90) [inline enforcer/src/main/java/org/wso2/choreo/connect/enforcer/api/APIFactory.java:74]: Lets reverse this condition, check if the type is ws, if not consider as a REST API

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go (hunk 4 above)
Comment (NomadXD) [reply adapter/internal/oasparser/envoyconf/routes_with_clusters.go:815]: For REST , there's a check to see whether resources are nil or not. 
https://github.com/wso2/product-microgateway/blob/6784b0c44b57aaaca842c1247398f0d2f50e7d87/adapter/internal/oasparser/envoyconf/routes_with_clusters.go#L818

For REST APIs if no resources, then the default will be `GET`. Is that an issue ?

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/AuthServer.java [hunk 6]
```diff
@@ -143,6 +144,7 @@ private static Server initServer() throws SSLException {
                 .workerEventLoopGroup(workerGroup)
                 .addService(ServerInterceptors.intercept(new ExtAuthService(), new AccessLogInterceptor()))
                 .addService(new HealthService())
+                .addService(new WebSocketFrameService())
```
Comment (Rajith90) [inline enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/AuthServer.java:147]: Shall we add the access log interceptor and also check whether it works

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/HttpRequestHandler.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (Rajith90) [inline enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/HttpRequestHandler.java:2]: Lets change the license year

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/WebSocketHandler.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+    /**
+     *
+     * @param webSocketFrameRequest - WebSocketFrameRequest received from WebSocketFrameRequest
+     * @return WebSocketResponseObject - Response sent via WebSocketFrameService
+     */
+    @Override
+    public WebSocketThrottleResponse process(WebSocketFrameRequest webSocketFrameRequest) {
+        WebSocketAPI matchedAPI = APIFactory.getInstance().getMatchedAPI(webSocketFrameRequest);
+        if (matchedAPI == null) {
```
Comment (Rajith90) [inline enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/WebSocketHandler.java:51]: What is the response retuned in this case. In rest API we create response object in here

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketResponseObserver.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    private void initializeThrottleKeys(WebSocketFrameRequest webSocketFrameRequest) {
+        Map<String, String> extAuthMetadata = webSocketFrameRequest.getMetadata().getExtAuthzMetadataMap();
+        String basePath = extAuthMetadata.get(APIConstants.GW_BASE_PATH_PARAM);
+        String version = extAuthMetadata.get(APIConstants.GW_VERSION_PARAM);
+        String applicationId = extAuthMetadata.get(MetadataConstants.APP_ID);
+        String streamId = extAuthMetadata.get(MetadataConstants.GRPC_STREAM_ID);
+        String apiContext = basePath + ':' + version;
+        this.apiThrottleKey = apiContext;
+        this.subscriptionThrottleKey = applicationId + ":" + apiContext;
+        this.applicationThrottleKey = applicationId;
```
Comment (Rajith90) [inline enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketResponseObserver.java:90]: application throttle key should have the user name as well

Comment (NomadXD) [reply enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketResponseObserver.java:90]: Will add the user name. This key is kept as a reference here. The application key used for throttling in defined in the WebSocketThrottleFilter correctly. https://github.com/wso2/product-microgateway/blob/6784b0c44b57aaaca842c1247398f0d2f50e7d87/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketThrottleFilter.java#L130

File: adapter/internal/api/apis_impl.go [hunk 10]
```diff
@@ -136,7 +136,8 @@ func extractAPIProject(payload []byte) (apiProject ProjectAPI, err error) {
...
+			!strings.Contains(file.Name, openAPIDir){
```
Comment (dinusha92) [inline adapter/internal/api/apis_impl.go:140]: Formatting issue.

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go (hunk 3 above)
Comment (NomadXD) [reply adapter/internal/oasparser/envoyconf/routes_with_clusters.go:845]: We can do that too. Shall I read from the config ? This value will decide how long the web socket connection is kept open. Setting it to 24 hrs means, the connection will be automatically closed after 24 hrs.

File: adapter/internal/oasparser/envoyconf/http_filters.go [hunk 11]
```diff
@@ -75,8 +82,20 @@ func getRouterHTTPFilter() *hcmv3.HttpFilter {
...
+	}
+	extAauth := getExtAuthzHTTPFilter()
+	mgwWebSocketWASM := getMgwWebSocketWASMFilter()
+	router := getRouterHTTPFilter()
+	upgradeFilters := []*hcmv3.HttpFilter{
+		cors,
+		extAauth,
+		mgwWebSocketWASM,
+		router,
+	}
```
Comment (dinusha92) [inline adapter/internal/oasparser/envoyconf/http_filters.go:97]: Shall we fix the formatting issues here.

Comment (dinusha92) [reply adapter/internal/oasparser/envoyconf/http_filters.go:97]: Ignore this comment.

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/WebSocketHandler.java (hunk 8 above)
Comment (NomadXD) [reply enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/WebSocketHandler.java:51]: `WebSocketThrottleResponse` contains the current throttle state and the UTC time in seconds for throttle out time.

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/api/APIFactory.java [hunk 12]
```diff
@@ -64,10 +66,18 @@ public void addApis(List<Api> apis) {
         ConcurrentHashMap<String, API> newApis = new ConcurrentHashMap<>();
 
         for (Api api : apis) {
-            RestAPI enforcerApi = new RestAPI();
-            enforcerApi.init(api);
-            String apiKey = getApiKey(enforcerApi);
-            newApis.put(apiKey, enforcerApi);
+            if (api.getApiType().equals(APIConstants.ApiType.REST)) {
```
Comment (dinusha92) [inline enforcer/src/main/java/org/wso2/choreo/connect/enforcer/api/APIFactory.java:69]: This should be the other way.
```suggestion
            if ((APIConstants.ApiType).REST.equals(api.getApiType())) {
```

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/HttpRequestHandler.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+                request.getAttributes().getRequest().getTime().getNanos() / 1000000;
+        String requestID = request.getAttributes().getRequest().getHttp().getId();
+        String address = "";
+        if (request.getAttributes().getSource().hasAddress() &&
+                request.getAttributes().getSource().getAddress().hasSocketAddress()) {
+            address = request.getAttributes().getSource().getAddress().getSocketAddress().getAddress();
+        }
+        address = FilterUtils.getClientIp(headers, address);
+        ResourceConfig resourceConfig = null;
+        if (api.getAPIConfig().getApiType().equals(APIConstants.ApiType.REST)) {
```
Comment (dinusha92) [inline enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/HttpRequestHandler.java:79]: Swap the check by putting the constant first.

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/WebSocketHandler.java (hunk 8 above)
Comment (dinusha92) [reply enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/WebSocketHandler.java:51]: Shouldn't we log a message if the matched API is null?

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketMetaDataFilter.java [hunk 14]
```diff
@@ -0,0 +1,89 @@
+package org.wso2.choreo.connect.enforcer.websocket;
```
Comment (dinusha92) [inline enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketMetaDataFilter.java:1]: Missing license header.

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketFrameContext.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
+package org.wso2.choreo.connect.enforcer.websocket;
```
Comment (dinusha92) [inline enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketFrameContext.java:1]: Missing license header

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/MetadataConstants.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
+package org.wso2.choreo.connect.enforcer.websocket;
```
Comment (dinusha92) [inline enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/MetadataConstants.java:1]: Missing license header.

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketThrottleFilter.java [hunk 17]
```diff
@@ -0,0 +1,274 @@
+package org.wso2.choreo.connect.enforcer.websocket;
```
Comment (dinusha92) [inline enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketThrottleFilter.java:1]: Missing license header

File: router/src/main/resources/Dockerfile [hunk 18]
```diff
@@ -24,6 +24,8 @@ ARG MG_USER_GROUP=wso2
 ARG MG_USER_GROUP_ID=802
 ARG MG_USER_HOME=/home/${MG_USER}
 
+#COPY maven/mgw-websocket.wasm /usr/local/bin/mgw-websocket.wasm
+#RUN chmod go+r /etc/envoy/envoy.yaml /usr/local/bin/mgw-websocket.wasm
```
Comment (dinusha92) [inline router/src/main/resources/Dockerfile:28]: Why have we commented this here? If this is not needed shall we remove this.

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go (hunk 4 above)
Comment (NomadXD) [reply adapter/internal/oasparser/envoyconf/routes_with_clusters.go]: fixed by adding a method to return resourceMethods

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/api/APIFactory.java (hunk 5 above)
Comment (NomadXD) [reply enforcer/src/main/java/org/wso2/choreo/connect/enforcer/api/APIFactory.java]: fixed

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/AuthServer.java (hunk 6 above)
Comment (NomadXD) [reply enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/AuthServer.java]: fixed

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/HttpRequestHandler.java (hunk 7 above)
Comment (NomadXD) [reply enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/HttpRequestHandler.java]: fixed

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/WebSocketHandler.java (hunk 8 above)
Comment (NomadXD) [reply enforcer/src/main/java/org/wso2/choreo/connect/enforcer/server/WebSocketHandler.java:51]: fixed

File: router/src/main/resources/Dockerfile (hunk 18 above)
Comment (NomadXD) [reply router/src/main/resources/Dockerfile]: fixed

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketThrottleFilter.java (hunk 17 above)
Comment (NomadXD) [reply enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketThrottleFilter.java:18]: fixed

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/MetadataConstants.java (hunk 16 above)
Comment (NomadXD) [reply enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/MetadataConstants.java:18]: fixed

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketFrameContext.java (hunk 15 above)
Comment (NomadXD) [reply enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketFrameContext.java:18]: fixed

File: enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketMetaDataFilter.java (hunk 14 above)
Comment (NomadXD) [reply enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketMetaDataFilter.java:18]: fixed

**Review by Rajith90 — APPROVED** [review]

**Pull request merged**